File: backend/internal/ws/messages.go
package ws

import "klopf-game/internal/game"

// MessageType represents the type of WebSocket message
type MessageType string

// Client -> Server message types
const (
	MsgCreateRoom     MessageType = "create_room"
	MsgJoinRoom       MessageType = "join_room"
	MsgReconnect      MessageType = "reconnect"
	MsgStartGame      MessageType = "start_game"
	MsgCloseRoom      MessageType = "close_room"
	MsgPlayCard       MessageType = "play_card"
	MsgKlopf          MessageType = "klopf"
	MsgKlopfResponse  MessageType = "klopf_response"
	MsgBlindDrei      MessageType = "blind_drei"
)

// Server -> Client message types
const (
	MsgRoomCreated        MessageType = "room_created"
	MsgRoomClosed         MessageType = "room_closed"
	MsgPlayerJoined       MessageType = "player_joined"
	MsgPlayerLeft         MessageType = "player_left"
	MsgGameStarted        MessageType = "game_started"
	MsgCardsDealt         MessageType = "cards_dealt"
	MsgCardPlayed         MessageType = "card_played"
	MsgKlopfInitiated     MessageType = "klopf_initiated"
	MsgKlopfResponseNeeded MessageType = "klopf_response_needed"
	MsgKlopfResolved      MessageType = "klopf_resolved"
	MsgTrickWon           MessageType = "trick_won"
	MsgRoundEnded         MessageType = "round_ended"
	MsgGameOver           MessageType = "game_over"
	MsgYourTurn           MessageType = "your_turn"
	MsgGameState          MessageType = "game_state"
	MsgError              MessageType = "error"
	MsgTimerUpdate        MessageType = "timer_update"
)

// BaseMessage is the base structure for all messages
type BaseMessage struct {
	Type MessageType `json:"type"`
}

// ClientMessage represents incoming messages from clients
type ClientMessage struct {
	Type       MessageType `json:"type"`
	PlayerName string      `json:"playerName,omitempty"`
	RoomCode   string      `json:"roomCode,omitempty"`
	PlayerID   string      `json:"playerId,omitempty"`
	CardID     string      `json:"cardId,omitempty"`
	Mitgehen   bool        `json:"mitgehen,omitempty"`
}

// ServerMessage represents outgoing messages to clients
type ServerMessage struct {
	Type     MessageType `json:"type"`
	RoomCode string      `json:"roomCode,omitempty"`
	PlayerID string      `json:"playerId,omitempty"`
	Player   *PlayerInfo `json:"player,omitempty"`
	State    *GameStateInfo `json:"state,omitempty"`
	Cards    []game.Card `json:"cards,omitempty"`
	Card     *game.Card  `json:"card,omitempty"`
	Level    int         `json:"level,omitempty"`
	WinnerID string      `json:"winnerId,omitempty"`
	Results  []RoundResult `json:"results,omitempty"`
	Error    string      `json:"error,omitempty"`
	TimeLeft int         `json:"timeLeft,omitempty"`
}

// PlayerInfo is the public info about a player
type PlayerInfo struct {
	ID        string `json:"id"`
	Name      string `json:"name"`
	Lives     int    `json:"lives"`
	CardCount int    `json:"cardCount"`
	Connected bool   `json:"connected"`
}

// GameStateInfo is the full game state sent to clients
type GameStateInfo struct {
	State           game.GameState `json:"state"`
	Players         []PlayerInfo   `json:"players"`
	CurrentPlayerID string         `json:"currentPlayerId"`
	TrickNumber     int            `json:"trickNumber"`
	RoundNumber     int            `json:"roundNumber"`
	CurrentTrick    *game.Trick    `json:"currentTrick,omitempty"`
	Klopf           *game.KlopfState `json:"klopf,omitempty"`
}

// RoundResult shows the outcome for a player in a round
type RoundResult struct {
	PlayerID   string `json:"playerId"`
	PlayerName string `json:"playerName"`
	LivesLost  int    `json:"livesLost"`
	LivesLeft  int    `json:"livesLeft"`
	IsLoser    bool   `json:"isLoser"`
}

// NewPlayerInfo creates a PlayerInfo from a game.Player
func NewPlayerInfo(p *game.Player) *PlayerInfo {
	return &PlayerInfo{
		ID:        p.ID,
		Name:      p.Name,
		Lives:     p.Lives,
		CardCount: len(p.Hand),
		Connected: p.Connected,
	}
}

// NewGameStateInfo creates a GameStateInfo from a game.Game
func NewGameStateInfo(g *game.Game) *GameStateInfo {
	players := make([]PlayerInfo, len(g.Players))
	for i, p := range g.Players {
		players[i] = *NewPlayerInfo(p)
	}

	return &GameStateInfo{
		State:           g.State,
		Players:         players,
		CurrentPlayerID: g.GetCurrentPlayerID(),
		TrickNumber:     g.TrickNumber,
		RoundNumber:     g.RoundNumber,
		CurrentTrick:    g.CurrentTrick,
		Klopf:           g.Klopf,
	}
}

File: packages/shared/src/messages.ts
import { Type, type Static } from '@sinclair/typebox';
import { CardSchema } from './card.js';
import { PlayerSchema } from './player.js';
import { GameStateInfoSchema, RoundResultSchema } from './game.js';

// ============================================
// Client -> Server Messages
// ============================================

export const CreateRoomMessage = Type.Object({
  type: Type.Literal('create_room'),
  playerName: Type.String(),
});

export const JoinRoomMessage = Type.Object({
  type: Type.Literal('join_room'),
  roomCode: Type.String(),
  playerName: Type.String(),
});

export const ReconnectMessage = Type.Object({
  type: Type.Literal('reconnect'),
  roomCode: Type.String(),
  playerId: Type.String(),
});

export const StartGameMessage = Type.Object({
  type: Type.Literal('start_game'),
});

export const CloseRoomMessage = Type.Object({
  type: Type.Literal('close_room'),
});

export const PlayCardMessage = Type.Object({
  type: Type.Literal('play_card'),
  cardId: Type.String(),
});

export const KlopfMessage = Type.Object({
  type: Type.Literal('klopf'),
});

export const KlopfResponseMessage = Type.Object({
  type: Type.Literal('klopf_response'),
  mitgehen: Type.Boolean(),
});

export const BlindDreiMessage = Type.Object({
  type: Type.Literal('blind_drei'),
});

export const SetStakesMessage = Type.Object({
  type: Type.Literal('set_stakes'),
  stakes: Type.Number(),
});

export const RequestRedealMessage = Type.Object({
  type: Type.Literal('request_redeal'),
});

export const RedealResponseMessage = Type.Object({
  type: Type.Literal('redeal_response'),
  agree: Type.Boolean(),
});

// Union of all client messages
export const ClientMessageSchema = Type.Union([
  CreateRoomMessage,
  JoinRoomMessage,
  ReconnectMessage,
  StartGameMessage,
  CloseRoomMessage,
  PlayCardMessage,
  KlopfMessage,
  KlopfResponseMessage,
  BlindDreiMessage,
  SetStakesMessage,
  RequestRedealMessage,
  RedealResponseMessage,
]);
export type ClientMessage = Static<typeof ClientMessageSchema>;

// ============================================
// Server -> Client Messages
// ============================================

export const RoomCreatedMessage = Type.Object({
  type: Type.Literal('room_created'),
  roomCode: Type.String(),
  playerId: Type.String(),
});

export const RoomClosedMessage = Type.Object({
  type: Type.Literal('room_closed'),
});

export const PlayerJoinedMessage = Type.Object({
  type: Type.Literal('player_joined'),
  player: PlayerSchema,
});

export const PlayerLeftMessage = Type.Object({
  type: Type.Literal('player_left'),
  playerId: Type.String(),
});

export const GameStartedMessage = Type.Object({
  type: Type.Literal('game_started'),
});

export const CardsDealtMessage = Type.Object({
  type: Type.Literal('cards_dealt'),
  cards: Type.Array(CardSchema),
});

export const CardPlayedMessage = Type.Object({
  type: Type.Literal('card_played'),
  playerId: Type.String(),
  card: CardSchema,
});

export const YourTurnMessage = Type.Object({
  type: Type.Literal('your_turn'),
});

export const KlopfInitiatedMessage = Type.Object({
  type: Type.Literal('klopf_initiated'),
  playerId: Type.String(),
  level: Type.Number(),
});

export const KlopfResponseNeededMessage = Type.Object({
  type: Type.Literal('klopf_response_needed'),
  level: Type.Number(),
});

export const KlopfResolvedMessage = Type.Object({
  type: Type.Literal('klopf_resolved'),
  level: Type.Number(),
});

export const TrickWonMessage = Type.Object({
  type: Type.Literal('trick_won'),
  winnerId: Type.String(),
});

export const RoundEndedMessage = Type.Object({
  type: Type.Literal('round_ended'),
  results: Type.Array(RoundResultSchema),
});

export const GameOverMessage = Type.Object({
  type: Type.Literal('game_over'),
  winnerId: Type.String(),
  perfectWin: Type.Boolean(),
  stakes: Type.Number(),
  winnings: Type.Number(),
});

export const GameStateMessage = Type.Object({
  type: Type.Literal('game_state'),
  state: GameStateInfoSchema,
});

export const ErrorMessage = Type.Object({
  type: Type.Literal('error'),
  error: Type.String(),
});

export const TimerUpdateMessage = Type.Object({
  type: Type.Literal('timer_update'),
  timeLeft: Type.Number(),
});

export const RedealRequestedMessage = Type.Object({
  type: Type.Literal('redeal_requested'),
  playerId: Type.String(),
});

export const RedealResponseNeededMessage = Type.Object({
  type: Type.Literal('redeal_response_needed'),
  redealCount: Type.Number(),
  maxRedeals: Type.Number(),
});

export const RedealPerformedMessage = Type.Object({
  type: Type.Literal('redeal_performed'),
  redealCount: Type.Number(),
  maxRedeals: Type.Number(),
});

export const RedealDeclinedMessage = Type.Object({
  type: Type.Literal('redeal_declined'),
});

// Union of all server messages
export const ServerMessageSchema = Type.Union([
  RoomCreatedMessage,
  RoomClosedMessage,
  PlayerJoinedMessage,
  PlayerLeftMessage,
  GameStartedMessage,
  CardsDealtMessage,
  CardPlayedMessage,
  YourTurnMessage,
  KlopfInitiatedMessage,
  KlopfResponseNeededMessage,
  KlopfResolvedMessage,
  TrickWonMessage,
  RoundEndedMessage,
  GameOverMessage,
  GameStateMessage,
  ErrorMessage,
  TimerUpdateMessage,
  RedealRequestedMessage,
  RedealResponseNeededMessage,
  RedealPerformedMessage,
  RedealDeclinedMessage,
]);
export type ServerMessage = Static<typeof ServerMessageSchema>;
